source/slang/slang-c-like-source-emitter.cpp

```diff
@@ -0,0 +1,5811 @@
+// slang-c-like-source-emitter.cpp
+#include "slang-c-like-source-emitter.h"
+
+#include "../core/slang-writer.h"
+#include "ir-bind-existentials.h"
+#include "ir-dce.h"
+#include "ir-entry-point-uniforms.h"
+#include "ir-glsl-legalize.h"
+
+#include "ir-link.h"
+#include "ir-restructure-scoping.h"
+#include "ir-specialize.h"
+#include "ir-specialize-resources.h"
+#include "ir-ssa.h"
+#include "ir-union.h"
+#include "ir-validate.h"
+#include "legalize-types.h"
+#include "lower-to-ir.h"
+#include "mangle.h"
+#include "name.h"
+#include "syntax.h"
+#include "type-layout.h"
+#include "visitor.h"
+
+#include "slang-source-stream.h"
+#include "slang-emit-context.h"
+#include "slang-mangled-lexer.h"
+
+#include <assert.h>
+
+namespace Slang {
+
+// represents a declarator for use in emitting types
+struct CLikeSourceEmitter::EDeclarator
```

Comment:
I've never seen this kind of declaration before. Is it actually possible to forward-declare a nested `struct` type and then put its definition out-of-line? If so, that's amazing and I wonder how I never learned about this before now.